internal/database/mysql_test.go

```diff
@@ -38,78 +37,134 @@ func setupDBAndOpenSubject(t *testing.T) *internalDatabase.MySQLDatabase {
+		log.Fatalln("Failed to use test_db -", err)
 	}
 
-	// Open the custom testing database with initial SQL setup
-	subject := internalDatabase.New("test_db", "root", password)
-	if subject == nil {
-		t.Fatal("Failed to open subject database")
-	}
+	// Open the internal database implementation to test
+	subject = New("test_db", "root", password)
 	if err := subject.Open(); err != nil {
 		t.Fatal("Failed to open subject database -", err)
 	}
+	return subject, probe
+}
+
+func teardown(t *testing.T, subject *MySQLDatabase, probe *sql.DB) {
+	t.Helper()
+
+	err := probe.Close()
+	if err != nil {
+		t.Fatal("Failed to close probe database connection -", err)
+	}
+	err = subject.Close()
+	if err != nil {
+		t.Fatal("Failed to close subject database connection -", err)
+	}
+}
 
-	return subject
+func TestConnections(t *testing.T) {
+	subject, probe := setup(t)
+	if err := subject.underlyingDB.Ping(); err != nil {
+		t.Fatal("Failed to open subject db properly -", err)
+	}
+	if err := probe.Ping(); err != nil {
+		t.Fatal("Failed to open probe db properly -", err)
+	}
+	teardown(t, subject, probe)
+	if err := subject.underlyingDB.Ping(); err == nil {
+		t.Fatal("Failed to close subject db properly -", err)
+	}
+	if err := probe.Ping(); err == nil {
+		t.Fatal("Failed to close probe db properly -", err)
+	}
 }
 
-func TestOpenAndClose(t *testing.T) {
-	subject := setupDBAndOpenSubject(t)
-	if subject == nil {
-		t.Error("Database connection is nil")
-		return
+func TestNew(t *testing.T) {
+	subject, probe := setup(t)
+	defer teardown(t, subject, probe)
+
+	expectedTables := map[string]map[string]struct{}{
+		"users":  {"id": {}, "username": {}, "email": {}, "password": {}, "created_on": {}},
+		"openid": {"id": {}, "created_on": {}},
 	}
-	// TODO(@Alishah634): Test that all the tables are generated.
-	err := subject.Close()
+
+	tables, err := probe.Query("SHOW TABLES FROM test_db")
 	if err != nil {
-		t.Error("Failed to close database -", err)
+		t.Fatal("Failed to query for all tables -", err)
+	}
+	for tables.Next() {
+		var table string
+		if err := tables.Scan(&table); err != nil || table == "" {
+			t.Fatal("Failed to scan tables -", err)
+		}
+		expectedCols, exists := expectedTables[table]
+		if !exists {
+			t.Fatal("Unexpected table:", table)
+		}
+		t.Run(
+			"verify_columns:"+table, func(t *testing.T) {
+				columns, err := probe.Query(fmt.Sprintf("SHOW COLUMNS FROM test_db.%s", table))
+				if err != nil {
+					t.Fatal("Failed to query for columns -", err)
+				}
+				for columns.Next() {
+					var field, typ, null, key, def, extra sql.NullString
+					err := columns.Scan(&field, &typ, &null, &key, &def, &extra)
+					if err != nil {
+						t.Fatal("Failed to scan column -", err)
+					}
+					delete(expectedCols, field.String)
+				}
+				if len(expectedCols) != 0 {
+					t.Log("Failed to create columns:")
+					for k := range expectedCols {
+						t.Log(k)
+					}
+					t.Fail()
+				}
+			},
+		)
+		delete(expectedTables, table)
```

Comment:
Lines 112 to 138, is there a more efficient solution?

Reply:
There is maybe a more readable solution, but any other sort of checking woudl require extra memory/ iterative checking at the end, so not really.